posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -0,0 +1,331 @@
++
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "empty.property")
+String emptyProperty; // Will throw a `DeploymentException` caused by a `NoSuchElementException`
+
+@Inject
+@ConfigProperty(name = "empty.array.property")
+String[] emptyArrayProperty; // Will throw a `DeploymentException` caused by a `NoSuchElementException`
+----
++
+and
++
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getValue("empty.property", String.class); // Will thorw a `NoSuchElementException`
+config.getValue("empty.array.property", String[].class); // Will thorw a `NoSuchElementException`
+----
++
+However these values can be retrieved "optionally":
++
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "empty.property")
+Optional<String> emptyProperty; // returns "Optional.empty"
+
+@Inject
+@ConfigProperty(name = "empty.array.property")
+Optional<String[]> emptyArrayProperty; // returns "Optional.empty"
+----
++
+and
++
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getOptionalValue("empty.property", String.class); // returns “Optional.empty"
+config.getOptionalValue("empty.array.property", String[].class); // returns “Optional.empty"
+----
+
+
+* This means that link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getValue-java.lang.String-java.lang.Class-[Config.getValue()] will never return null; a `NoSuchElementException` will be thrown if the property:
+
+** Is not defined
+** Is defined as an empty String (`""`)
+** Is converted to `null` (considered to be "empty") by its Converter
+
+[#Expanding-Config-API]
+=== Expanding the Config API
+Two new methods have been added to the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html[Config API class]:
+
+1. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getValues-java.lang.String-java.lang.Class-[Config.getValues()]: `List<T> Config.getValues(String propertyName, Class<T> propertyType)`
+
+2. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/Config.html#getOptionalValues-java.lang.String-java.lang.Class-[Config.getOptionalValues()]: `Optional <List<T>>Config.getOptionalValues(String propertyName, Class<T> propertyType)`
+
+The methods have been added to enable you to retrieve multi-valued config property values as lists instead of arrays. The methods return the resolved property values with the specified `propertyType` for the specified `propertyName`. For example:
+
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getOptionalValues("ports", Integer.class) // returns "Optional[[9080, 9081, 9082]]"
```

Comment:
even if we ignore the quotes, i'm not sure what `Optional[[9080, 9081, 9082]]` really is just from looking at it ... need to be clearer that the return type is `Optional<List<Integer>>`

Reply:
I think the square brackets implies it's a `List` though, right?